picard/webservice/__init__.py

```diff
@@ -65,6 +66,136 @@ def __init__(self, response_type):
         super().__init__(message)
 
 
+class WSRequest(QtNetwork.QNetworkRequest):
+    """Represents a single HTTP request.
+
+    :param method: HTTP method.  One of ``GET``, ``POST``, ``PUT``, or ``DELETE``.
+    :param host: Hostname.
+    :param port: TCP port number (80 or 443).
+    :param path: Path component.
+    :param handler: Callback which takes a 3-tuple of `(str:document,
+        QNetworkReply:reply, QNetworkReply.Error:error)`.
+    :param parse_response_type: Specifies that request either sends or accepts
+        data as ``application/{{response_mimetype}}``.
+    :param data: Data to include with ``PUT`` or ``POST`` requests.
+    :param mblogin: Hints that this request should be tied to a MusicBrainz
+        account, requiring that we obtain an OAuth token first.
+    :param cacheloadcontrol: See `QNetworkRequest.CacheLoadControlAttribute`.
+    :param refresh: Indicates a user-specified resource refresh, such as when
+        the user wishes to reload a release.  Marks the request as high priority
+        and disables caching.
+    :param access_token: OAuth token.
+    :param queryargs: `dict` of query arguments.
+    :param retries: Current retry attempt number.
+    :param priority: Indicates that this is a high priority request.  (See
+        `WebService._run_next_task`.)
+    :param important: Indicates that this is an important request.  (Ditto.)
+    """
+
+    def __init__(self, method, host, port, path, handler, parse_response_type=None, data=None,
+                 mblogin=False, cacheloadcontrol=None, refresh=None,
+                 queryargs=None, priority=False, important=False):
+        url = build_qurl(host, port, path=path, queryargs=queryargs)
+        super().__init__(url)
+
+        # These two are codependent (see _update_authorization_header) and must
+        # be initialized explicitly.
+        self._access_token = None
+        self._mblogin = None
+
+        self._retries = 0
+
+        self.method = method
+        self.host = host
+        self.port = port
+        self.path = path
+        self.handler = handler
+        self.parse_response_type = parse_response_type
+        self.response_parser = None
+        self.response_mimetype = None
+        self.data = data
+        self.mblogin = mblogin
+        self.cacheloadcontrol = cacheloadcontrol
+        self.refresh = refresh
+        self.queryargs = queryargs
+        self.priority = priority
+        self.important = important
+
+        self.access_token = None
+        self._init_headers()
+
+    def _init_headers(self):
+        self.setRawHeader(b"User-Agent", USER_AGENT_STRING.encode('utf-8'))
```

Comment:
Not sure it makes any difference but there's a known header for this (`QNetworkRequest.UserAgentHeader`), but better to use it (replacing `setRawHeader()` with `setHeader()`)